Svc/SeqDispatcher/SeqDispatcher.cpp

```diff
@@ -0,0 +1,199 @@
+      return i;
+    }
+  }
+  return -1;
+}
+
+bool SeqDispatcher::runSequence(FwIndexType sequencerIdx,
+                                const Fw::StringBase& fileName,
+                                Fw::Wait block) {
+  FW_ASSERT(sequencerIdx >= 0 && sequencerIdx < SeqDispatcherSequencerPorts,
+            sequencerIdx);
+
+  // if the seqRunOut port at this idx isn't connected
+  if (!this->isConnected_seqRunOut_OutputPort(sequencerIdx)) {
+    this->log_WARNING_HI_InvalidSequencer(static_cast<U16>(sequencerIdx));
+    return false;
+  }
+
+  // if the sequencer is not available to run a new sequence
+  if (this->m_entryTable[sequencerIdx].state !=
+      SeqDispatcher_CmdSequencerState::AVAILABLE) {
+    this->log_WARNING_HI_InvalidSequencer(static_cast<U16>(sequencerIdx));
+    return false;
+  }
+
+  // ok, good to use this sequencer
+  if (block == Fw::Wait::NO_WAIT) {
+    this->m_entryTable[sequencerIdx].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK;
+  } else {
+    this->m_entryTable[sequencerIdx].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_BLOCK;
+  }
+
+  this->m_sequencersAvailable--;
+  this->tlmWrite_sequencersAvailable(this->m_sequencersAvailable);
+  this->m_entryTable[sequencerIdx].sequenceRunning = fileName;
+
+  this->m_dispatchedCount++;
+  this->tlmWrite_dispatchedCount(this->m_dispatchedCount);
+  this->seqRunOut_out(sequencerIdx,
+                      this->m_entryTable[sequencerIdx].sequenceRunning);
+
+  return true;
+}
+
+void SeqDispatcher::seqStartIn_handler(
+    NATIVE_INT_TYPE portNum, //!< The port number
+    const Fw::StringBase& fileName //!< The sequence file name
+) {
+  FW_ASSERT(portNum >= 0 && portNum < SeqDispatcherSequencerPorts, portNum);
+  if (this->m_entryTable[portNum].state ==
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_BLOCK ||
+      this->m_entryTable[portNum].state ==
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK) {
+    // we were aware of this sequencer running a sequence
+    if (this->m_entryTable[portNum].sequenceRunning != fileName) {
+      // uh oh. m_sequencesRunning is wrong
+      // let's just update it to be correct. nothing we can do about
+      // it except raise a warning and update our state
+      this->log_WARNING_HI_UnexpectedSequenceStarted(static_cast<U16>(portNum), fileName, this->m_entryTable[portNum].sequenceRunning);
+      this->m_entryTable[portNum].sequenceRunning = fileName;
+    }
+  } else {
+    // we were not aware that this sequencer was running. ground must have
+    // directly commanded that specific sequencer
+
+    // update the state
+    this->m_entryTable[portNum].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK;
+    this->m_entryTable[portNum].sequenceRunning = fileName;
+    this->m_sequencersAvailable--;
+    this->tlmWrite_sequencersAvailable(this->m_sequencersAvailable);
+  }
```

Comment:
If a Sequence is in Available state and you receive a Start, that seems to be an anomaly.  I would suggest sending an EVR Warning instead of just silently transitioning to a NonBlock state.

Reply:
I agree with this, can we fix?

Reply:
Yes, will fix.